### PR TITLE
codet API

### PR DESCRIPTION
**Pull request by kroening**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/std_code.h [hunk 1]
```diff
@@ -48,6 +48,26 @@ class codet:public exprt
     set_statement(statement);
   }
 
+  /// \param statement: Specifies the type of the `codet` to be constructed,
+  ///   e.g. `ID_block` for a \ref code_blockt or `ID_assign` for a
+  ///   \ref code_assignt.
+  explicit codet(const irep_idt &statement, operandst &&_op)
+    : exprt(ID_code, empty_typet())
+  {
+    set_statement(statement);
```
Comment (tautschnig) [inline src/util/std_code.h]: How about using a delegating constructor rather than duplicating all the work?

File: src/util/std_code.h [hunk 2]
```diff
@@ -48,6 +48,26 @@ class codet:public exprt
...
+    operands() = std::move(_op);
+  }
+
+  /// \param statement: Specifies the type of the `codet` to be constructed,
+  ///   e.g. `ID_block` for a \ref code_blockt or `ID_assign` for a
+  ///   \ref code_assignt.
+  explicit codet(const irep_idt &statement, const operandst &_op)
+    : exprt(ID_code, empty_typet())
+  {
+    set_statement(statement);
```
Comment (tautschnig) [inline src/util/std_code.h]: As above, use a delegating constructor.

File: src/util/std_code.h [hunk 3]
```diff
@@ -48,6 +48,26 @@ class codet:public exprt
...
+    operands() = std::move(_op);
+  }
+
+  /// \param statement: Specifies the type of the `codet` to be constructed,
+  ///   e.g. `ID_block` for a \ref code_blockt or `ID_assign` for a
+  ///   \ref code_assignt.
+  explicit codet(const irep_idt &statement, const operandst &_op)
```
Comment (tautschnig) [inline src/util/std_code.h]: Having done a bunch of profiling and studied the profiles: we can likely get away with a single constructor taking `operandst` by value, and then `std::move` it in place.

**Review by tautschnig — APPROVED** [review]

File: src/util/std_code.h (hunk 1 above)
Comment (kroening) [reply src/util/std_code.h]: ok

File: src/util/std_code.h (hunk 2 above)
Comment (kroening) [reply src/util/std_code.h]: ok

File: src/util/std_code.h (hunk 3 above)
Comment (kroening) [reply src/util/std_code.h]: I suspected as much, and like the reduction in noise.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 9a5345ea5bf95b5cf620dbf92da96e52a725c6ad).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99980171
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by tautschnig** [comment]
@peterschrammel @svorenova @thk123 I suspect the last commit in this PR breaks the TG build. Could a bump be created, please?

**Review by smowton — APPROVED** [review]

**Comment by antlechner** [comment]
We have a passing TG bump for this now. :+1:

**Comment by tautschnig** [comment]
@antlechner Depending on how far you went with the changes, are those actually mergeable even without the changes in this PR? Might be nice to reduce dependencies/breakage.

**Review by peterschrammel — APPROVED** [review]

**Comment by antlechner** [comment]
@tautschnig I think they should be, I'm currently waiting for it to (hopefully) pass with just the code changes in TG, and without the actual bump. So if you don't mind waiting just a bit longer before merging this PR, that would be great, but as far as I know it's not a requirement yet (we only started doing those code changes in anticipation of a bump without the actual bump very recently I think). I'll post another update here soon.

**Comment by antlechner** [comment]
@tautschnig The TG changes are merged, so this PR can be merged without causing any temporary incompatibilities.

**Pull request merged**